File: java/integTest/src/test/java/glide/ErrorHandlingTests.java
/** Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0 */
package glide;

import static glide.TestUtilities.commonClientConfig;
import static org.junit.jupiter.api.Assertions.assertAll;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import glide.api.GlideClient;
import glide.api.models.configuration.GlideClientConfiguration;
import glide.api.models.configuration.NodeAddress;
import glide.api.models.exceptions.ClosingException;
import glide.api.models.exceptions.RequestException;
import java.net.ServerSocket;
import java.util.concurrent.ExecutionException;
import lombok.SneakyThrows;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.Timeout;

@Timeout(10) // seconds
public class ErrorHandlingTests {

    @Test
    @SneakyThrows
    public void basic_client_tries_to_connect_to_wrong_address() {
        var exception =
                assertThrows(
                        ExecutionException.class,
                        () ->
                                GlideClient.createClient(
                                                GlideClientConfiguration.builder()
                                                        .address(NodeAddress.builder().port(getFreePort()).build())
                                                        .build())
                                        .get());
        assertAll(
                () -> assertTrue(exception.getCause() instanceof ClosingException),
                () -> assertTrue(exception.getCause().getMessage().contains("Connection refused")));
    }

    @Test
    @SneakyThrows
    public void basic_client_tries_wrong_command() {
        try (var regularClient = GlideClient.createClient(commonClientConfig().build()).get()) {
            var exception =
                    assertThrows(
                            ExecutionException.class,
                            () -> regularClient.customCommand(new String[] {"pewpew"}).get());
            assertAll(
                    () -> assertTrue(exception.getCause() instanceof RequestException),
                    () -> assertTrue(exception.getCause().getMessage().contains("unknown command")));
        }
    }

    @Test
    @SneakyThrows
    public void basic_client_tries_wrong_command_arguments() {
        try (var regularClient = GlideClient.createClient(commonClientConfig().build()).get()) {
            var exception =
                    assertThrows(
                            ExecutionException.class,
                            () -> regularClient.customCommand(new String[] {"ping", "pang", "pong"}).get());
            assertAll(
                    () -> assertTrue(exception.getCause() instanceof RequestException),
                    () ->
                            assertTrue(exception.getCause().getMessage().contains("wrong number of arguments")));
        }
    }

    @SneakyThrows
    private int getFreePort() {
        try (ServerSocket s = new ServerSocket(0)) {
            return s.getLocalPort();
        }
    }
}

File: csharp/tests/Integration/ErrorHandlingTests.cs
// Copyright Valkey GLIDE Project Contributors - SPDX Identifier: Apache-2.0

using static Glide.ConnectionConfiguration;
using static Glide.Errors;

using TimeoutException = Glide.Errors.TimeoutException;

namespace Tests.Integration;

public class ErrorHandlingTests
{
    [Fact]
    public async Task ErrorIfConnectionFailed() =>
        await Assert.ThrowsAsync<ConnectionException>(async () =>
            await GlideClient.CreateClient(new StandaloneClientConfigurationBuilder().WithAddress(null, 42).Build())
        );

    [Fact(Skip = "Deactivated until #3395 merged")]
    public async Task ErrorIfTimedOut() =>
        await Assert.ThrowsAsync<TimeoutException>(async () =>
            await TestConfiguration.DefaultClusterClient().CustomCommand(["debug", "sleep", "1"])
        );

    [Fact]
    public async Task ErrorIfIncorrectArgs() =>
        await Assert.ThrowsAsync<RequestException>(async () =>
            await TestConfiguration.DefaultStandaloneClient().CustomCommand(["ping", "pong", "pang"])
        );
}
